### PR TITLE
Generated Docs Pipeline

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
This PR introduces a BuildKite pipeline for producing generated documentation and publishing it to Readme. This PR implements the first part of this. (See the TODO for the second part).

You can see an example build here - https://buildkite.com/improbable/gdk-for-unity-generate-docs/builds/9#_ where the generated docs are present in the artifacts.

### CODE REVIEW

File: .buildkite/generate-docs.definition.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+agent_queue_id: trigger-pipelines
+description: Generate API docs for Readme.io
```
Comment (paulbalaji) [inline .buildkite/generate-docs.definition.yaml]:
```suggestion
description: Generate API docs for ReadMe.io
```
?

File: .buildkite/generate-docs.steps.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  retry:
+    automatic:
+        # These are designed to trap and retry failures because agent lost connection. Agent exits with -1 in this case.
+      - exit_status: -1
+        limit: 3
+      - exit_status: 255
+        limit: 3
+      - exit_status: 240 # Within platform, we interpret this to be a Docker-related exit code
+        limit: 3
+
```
Comment (paulbalaji) [inline .buildkite/generate-docs.steps.yaml]: extra line

File: ci/generate-docs.sh [hunk 3]
```diff
@@ -19,58 +20,37 @@ function cleanUp() {
...
+    -v "${CODE_DIR}/workers/unity/Packages:/input" \
+    -v "${OUTPUT_DIR}:/output" \
+    docgen \
+        --target-namespace="Improbable.Gdk" \
+        --namespace-filter=".*EditmodeTests" \
+        --namespace-filter=".*DeploymentLauncher" \
+        --namespace-filter=".*PlaymodeTests" \
+        --git-tag="${TAG}"
+
+# TODO: Trigger Filip's pipeline when that's live
```
Comment (paulbalaji) [inline ci/generate-docs.sh]: What does Filip's pipeline do?

File: ci/generate-docs.sh [hunk 4]
```diff
@@ -2,12 +2,13 @@
...
-    exit 1
-fi
+cd "$(dirname "$0")/../"
 
-TAG="${1}"
+CURRENT_DIR=$(pwd)
+TMP_DIR=$(mktemp -d)
+OUTPUT_DIR="${CURRENT_DIR}/docs-output"
+rm -rf "${OUTPUT_DIR}"
+mkdir -p "${OUTPUT_DIR}"
```
Comment (zeroZshadow) [inline ci/generate-docs.sh]: no need for the `-p`, since the directory was just removed

File: ci/generate-docs.sh [hunk 5]
```diff
@@ -19,58 +20,37 @@ function cleanUp() {
...
 CODE_DIR="${TMP_DIR}/code"
-DOCS_DIR="${TMP_DIR}/docs"
 
-# Clone docgen repo.
-git clone "${DOCGEN_CLONE_URL}" "${DOCGEN_DIR}"
-
-# Clone and checkout correct code tag.
 git clone "${CLONE_URL}" "${CODE_DIR}"
 pushd "${CODE_DIR}"
     git checkout "${TAG}"
```
Comment (zeroZshadow) [inline ci/generate-docs.sh]: Since this is a tag. This could use `git clone url --branch <tag> --depth1` ?

File: ci/generate-docs.sh [hunk 6]
```diff
@@ -19,58 +20,37 @@ function cleanUp() {
...
-    --output-directory="${DOCS_DIR}/docs/api" \
-    --namespace-filter=".*EditmodeTests" \
-    --namespace-filter=".*DeploymentLauncher" \
-    --namespace-filter=".*PlaymodeTests" \
-    --api-path="api" \
-    --git-tag="${TAG}"
+# Clone docgen repo.
+DOCGEN_CLONE_URL="git@github.com:improbable/gdk-for-unity-docgen.git"
+DOCGEN_DIR="${TMP_DIR}/docgen"
+git clone "${DOCGEN_CLONE_URL}" "${DOCGEN_DIR}"
```
Comment (zeroZshadow) [inline ci/generate-docs.sh]: No pinning?

File: ci/generate-docs.sh (hunk 3 above)
Comment (jamiebrynes7) [reply ci/generate-docs.sh]: Uploads the docs given a build where they are in the artifacts

File: ci/generate-docs.sh (hunk 6 above)
Comment (jamiebrynes7) [reply ci/generate-docs.sh]: Not at the moment, I'm going to setup a docker image publish pipeline on the docgen repo so we can pin the image tag.

Soon ™️

**Pull request merged**